File: src/services/audit/detectors/ad/gpo/orphaned.ts
/**
 * GPO_ORPHANED - Orphaned GPO (Phase 4)
 */

import { Finding } from '../../../../../types/finding.types';
import { ADGPO, GPOLink } from '../../../../../types/gpo.types';

/**
 * Detect orphaned GPOs
 *
 * GPOs that exist in AD but have missing SYSVOL content, or vice versa.
 *
 * @param gpos - Array of GPOs
 * @param _links - Array of GPO links (not used)
 * @param includeDetails - Whether to include affected entity details
 * @returns Finding for GPO_ORPHANED
 */
export function detectGpoOrphaned(
  gpos: ADGPO[],
  _links: GPOLink[],
  includeDetails: boolean
): Finding {
  // Check for GPOs with potential orphan indicators
  // A proper check would compare AD GPOs vs SYSVOL folders
  const affected = gpos.filter((gpo) => {
    // Missing SYSVOL path indicates potential orphan
    const hasSysvolPath = gpo.gPCFileSysPath && gpo.gPCFileSysPath.length > 0;
    // Missing version might indicate corruption
    const hasVersion = gpo.versionNumber !== undefined && gpo.versionNumber > 0;
    // Check for obviously broken GPOs
    const hasName = gpo.displayName || gpo.cn;

    return !hasSysvolPath || !hasVersion || !hasName;
  });

  return {
    type: 'GPO_ORPHANED',
    severity: 'medium',
    category: 'gpo',
    title: 'Potentially Orphaned GPOs',
    description:
      'GPOs that may be orphaned (missing SYSVOL content or AD object). ' +
      'Orphaned GPOs can cause processing errors and may indicate tampering.',
    count: affected.length,
    affectedEntities: includeDetails
      ? affected.map((g) => g.displayName || g.cn || g.dn)
      : undefined,
    details: {
      recommendation:
        'Compare AD GPOs with SYSVOL folders. Use gpotool.exe or Get-GPO to identify orphans. ' +
        'Delete orphaned GPOs after verification.',
    },
  };
}

File: internal/audit/detectors/ad/gpo/orphaned.go
package gpo

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// OrphanedDetector checks for orphaned GPOs
type OrphanedDetector struct {
	audit.BaseDetector
}

// NewOrphanedDetector creates a new detector
func NewOrphanedDetector() *OrphanedDetector {
	return &OrphanedDetector{
		BaseDetector: audit.NewBaseDetector("GPO_ORPHANED", audit.CategoryGPO),
	}
}

// Detect executes the detection
func (d *OrphanedDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	var affected []string

	for _, gpo := range data.GPOs {
		// Check for GPOs with potential orphan indicators
		hasSysvolPath := gpo.FilePath != ""
		hasName := gpo.DisplayName != "" || gpo.CN != ""

		if !hasSysvolPath || !hasName {
			name := gpo.DisplayName
			if name == "" {
				name = gpo.CN
			}
			if name == "" {
				name = gpo.DistinguishedName
			}
			affected = append(affected, name)
		}
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "Potentially Orphaned GPOs",
		Description: "GPOs that may be orphaned (missing SYSVOL content or AD object). Orphaned GPOs can cause processing errors and may indicate tampering.",
		Count:       len(affected),
		Details: map[string]interface{}{
			"recommendation": "Compare AD GPOs with SYSVOL folders. Use gpotool.exe or Get-GPO to identify orphans. Delete orphaned GPOs after verification.",
		},
	}

	if len(affected) > 0 {
		finding.AffectedEntities = toAffectedGPOEntities(affected)
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewOrphanedDetector())
}
